src/node/desktop/src/ui/widgets/choose-r.ts

```diff
@@ -14,25 +14,86 @@
  */
 
 import { dialog, ipcMain } from 'electron';
-import { findDefault32Bit, findDefault64Bit } from '../../main/detect-r';
+import { detectREnvironment, findDefault32Bit, findDefault64Bit } from '../../main/detect-r';
 import { logger } from '../../core/logger';
 import { ModalDialog } from '../modal-dialog';
 
 import { initI18n } from '../../main/i18n-manager';
-import i18next from 'i18next';
+import i18next, { t } from 'i18next';
 import { CallbackData } from './choose-r/preload';
 import { ElectronDesktopOptions } from '../../main/preferences/electron-desktop-options';
 
 declare const CHOOSE_R_WEBPACK_ENTRY: string;
 declare const CHOOSE_R_PRELOAD_WEBPACK_ENTRY: string;
 
+function checkValid(data: CallbackData) {
+
+  // get path to R
+  const rBinaryPath = data.binaryPath as string;
+
+  // try to run it
+  const [rEnvironment, err] = detectREnvironment(rBinaryPath);
+  if (err) {
+    
+    const response = dialog.showMessageBoxSync({
+      title: t('chooseRDialog.rLaunchFailedTitle'),
+      message: t('chooseRDialog.rLaunchFailedMessage'),
+      buttons: [
+        t('common.buttonOk'),
+        t('common.buttonCancel')
+      ],
+    });
+
+    return response === 0;
+
+  }
+
+  logger().logDebug(`Validated R: ${rEnvironment.rScriptPath}`);
+  return true;
+
+}
+
 export class ChooseRModalWindow extends ModalDialog<CallbackData | null> {
   private rInstalls: string[];
 
+  private channels: string[] = [
```

Comment:
I feel like we need a more disciplined way of implementing IPC handlers, that could be shared across other windows doing this sort of communication.

Reply:
Agree